server/pulp/server/content/sources/container.py

```diff
@@ -189,28 +156,149 @@ def download_failed(self, report):
+        """
+        :param primary: The *primary* content source used when requested content
+            download requests cannot be satisfied using alternate content sources.
+        :type primary: PrimarySource
+        :param container: A content container.
+        :type container: ContentContainer
+        :param requests: An iterable of: pulp.server.content.sources.model.Request.
+        :type requests: iterable
+        :param listener: An optional download request listener.
+        :type listener: Listener
+        """
+        self.primary = primary
+        self.container = container
+        self.requests = requests
+        self.listener = listener
+
+    @property
+    def sources(self):
+        """
+        The list of available sources.
+
+        :return: The list of available sources.
+        :rtype: list
+        """
+        return self.container.sources
+
+    def __call__(self):
+        """
+        Begin processing the batch of requests.
+        Download files using available alternate content sources.
+        An attempt is made to satisfy each download request using the alternate
+        content sources in the order specified by priority.  The specified
+        downloader is designated as the primary source and is used in the event that
+        the request cannot be completed using alternate sources.
+
+        :return: The download report.
+        :rtype: DownloadReport
+        """
+        raise NotImplementedError()
+
+
+class Serial(Batch):
+    """
+    Provides sequential batch processing of a collection of content download requests.
+    This approach does *not* use threading.
+    """
+
+    def __call__(self):
+        """
+        Begin processing the batch of requests.
+        Download files using available alternate content sources.
+        An attempt is made to satisfy each download request using the alternate
+        content sources in the order specified by priority.  The specified
+        downloader is designated as the primary source and is used in the event that
+        the request cannot be completed using alternate sources.
+
+        :return: The download report.
+        :rtype: DownloadReport
+        """
+        report = DownloadReport()
+        report.total_sources = len(self.sources)
+        for request in self.requests:
+            event = Started(request)
+            event(self.listener)
+            request.find_sources(self.primary, self.sources)
+            for source, url in request.sources:
+                details = report.downloads.setdefault(source.id, DownloadDetails())
+                try:
+                    self._download(url, request.destination, source)
+                    details.total_succeeded += 1
+                    request.downloaded = True
+                    event = Succeeded(request)
+                    event(self.listener)
+                    break
+                except DownloadFailed, df:
```

Comment:
We should start using ``except ... as`` for future compatibility. The comma notation doesn't work in python 3.